File: 0076-minimum-window-substring/0076-minimum-window-substring.js
function minWindow(s, t) {
    if (!s || !t || s.length < t.length) return "";

    // Fixed Int32 array to track ASCII character counts efficiently
    const targetCounts = new Int32Array(128);
    const windowCounts = new Int32Array(128);

    let requiredMatches = 0;
    for (let i = 0; i < t.length; i++) {
        const ch = t.charCodeAt(i);
        if (targetCounts[ch] === 0) requiredMatches++;
        targetCounts[ch]++;
    }

    let left = 0, right = 0, formedMatches = 0;
    let minLen = Infinity, minLeftStart = 0;

    while (right < s.length) {
        const rightChar = s.charCodeAt(right);
        windowCounts[rightChar]++;

        if (targetCounts[rightChar] > 0 && windowCounts[rightChar] === targetCounts[rightChar]) {
            formedMatches++;
        }

        while (formedMatches === requiredMatches) {
            const currentLen = right - left + 1;
            if (currentLen < minLen) {
                minLen = currentLen;
                minLeftStart = left;
            }

            const leftChar = s.charCodeAt(left);
            windowCounts[leftChar]--;

            if (targetCounts[leftChar] > 0 && windowCounts[leftChar] < targetCounts[leftChar]) {
                formedMatches--;
            }
            left++;
        }
        right++;
    }

    return minLen === Infinity ? "" : s.substring(minLeftStart, minLeftStart + minLen);
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (s == null || t == null || s.length() < t.length()) {
            return "";
        }

        // ASCII frequency tracking arrays to skip HashMap instantiation overhead
        int[] targetCounts = new int[128];
        int[] windowCounts = new int[128];

        // Step 1: Populate target frequencies from string t
        int requiredMatches = 0;
        for (int i = 0; i < t.length(); i++) {
            char ch = t.charAt(i);
            if (targetCounts[ch] == 0) {
                requiredMatches++; // Count unique characters to satisfy
            }
            targetCounts[ch]++;
        }

        int left = 0;
        int right = 0;
        int formedMatches = 0;

        // Variables tracking our optimal result window
        int minLen = Integer.MAX_VALUE;
        int minLeftStart = 0;

        // Step 2: Expand the right side of the window
        while (right < s.length()) {
            char rightChar = s.charAt(right);
            windowCounts[rightChar]++;

            // If the character count matches its exact requirement in t, increment match metric
            if (targetCounts[rightChar] > 0 && windowCounts[rightChar] == targetCounts[rightChar]) {
                formedMatches++;
            }

            // Step 3: Contract the window from the left once all requirements are met
            while (formedMatches == requiredMatches) {
                int currentWindowLen = right - left + 1;
                
                // Track the absolute smallest valid window coordinates
                if (currentWindowLen < minLen) {
                    minLen = currentWindowLen;
                    minLeftStart = left;
                }

                char leftChar = s.charAt(left);
                windowCounts[leftChar]--;

                // If dropping this character breaks our match requirements, exit contraction loop
                if (targetCounts[leftChar] > 0 && windowCounts[leftChar] < targetCounts[leftChar]) {
                    formedMatches--;
                }
                
                left++; // Shrink window bounds forward
            }
            
            right++; // Expand window bounds forward
        }

        return minLen == Integer.MAX_VALUE ? "" : s.substring(minLeftStart, minLeftStart + minLen);
    }
}
